In [ ]:
#Importing libraries
import numpy as np
import pandas as pd 
import seaborn as sns
import matplotlib.pyplot as plt
import datetime as dt
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error 
from math import sqrt

In [ ]:
#Importing the data 
data_raw = pd.read_csv("dataset_1.csv", header= 0)
data = data_raw

#Converting the Week column into Number of weeks
def createList(a, b): 
    return [item for item in range(a, b+1)] 
a, b = 1, len(data.Week)

data["Week_Num"] = createList(a, b)

In [ ]:
#Remaining the column name
data.columns= ["Week", "Count", "Week_Num"]
#Dropping Week column and storing it into new variable
data_2d = data.iloc[:,1:]

final_result = []

#### Converting the data from 1 dimanetional to n dimentional. In our case the value of n = 2

In [ ]:
#Using lag method to increase the dimensionality
data_2d["Count_lag1"] = data_2d.Count.shift(1)

#Excluding the top row with Nan values to make data clean
data_2d = data_2d.iloc[1:]

#Assigning x and y values
x = data_2d.iloc[:,1:]
y = data_2d.iloc[:, 0]

**Note: According to sklearn library in knn classifier if we allocate the power parameter for default metric Minkowski as p = 1, the algorithm calculate the values using manhattan distance.

In [ ]:
#Applying and fitting the knn model for n = 2
def knn_regressor(k_value, distance_metric, x, y):
  model_2d = KNeighborsRegressor(n_neighbors = k_value, algorithm='auto', leaf_size=30, metric='minkowski', p = distance_metric)
  model_2d.fit(x, y)
  return model_2d

#Forming the loop to find all the combination of K-values and Distance metrics
k_values = [1,3,5,7]
distance_metric = [1,2]

for i in k_values:
  print("\nVALUE OF k", i)
  for k in distance_metric:
    model_2d = knn_regressor(i,k, x, y)
    #Calculating performance on training set
    if k == 1:
      print('Value of distance_metric: Manhattan')
    else:
      print('Value of distance_metric: Euclidean')

    prediction = model_2d.predict(x)
    error = sqrt(mean_squared_error(y,prediction))
    print('Root Mean Square Error (n = 2):',error)
    
    prediction_last_week = model_2d.predict([[259,37]])
    final_result.append(['2',i,k,error,round(prediction_last_week[0])])


VALUE OF k 1
Value of distance_metric: Manhattan
Root Mean Square Error (n = 2): 0.0
Value of distance_metric: Euclidean
Root Mean Square Error (n = 2): 0.0

VALUE OF k 3
Value of distance_metric: Manhattan
Root Mean Square Error (n = 2): 8.92342290343357
Value of distance_metric: Euclidean
Root Mean Square Error (n = 2): 9.029166063002808

VALUE OF k 5
Value of distance_metric: Manhattan
Root Mean Square Error (n = 2): 9.841409113197832
Value of distance_metric: Euclidean
Root Mean Square Error (n = 2): 9.791880832228205

VALUE OF k 7
Value of distance_metric: Manhattan
Root Mean Square Error (n = 2): 10.448478398239379
Value of distance_metric: Euclidean
Root Mean Square Error (n = 2): 10.324416543595573


#### Converting the data from 1 dimanetional to n dimentional. In our case the value of n = 3

In [ ]:
#Using lag method to increase the dimensionality
data_3d = data.iloc[:,1:]
data_3d["Count_lag1"] = data_3d.Count.shift(1)
data_3d["Count_lag2"] = data_3d.Count.shift(2)

#Excluding the top row with Nan values to make data clean
data_3d = data_3d.iloc[2:]

#Assigning x and y values
x3 = data_3d.iloc[:,1:]
y3 = data_3d.iloc[:, 0]

**Note: According to sklearn library in knn classifier if we allocate the power parameter for default metric Minkowski as p = 1, the algorithm calculate the values using manhattan distance.

In [ ]:
#Applying and fitting the knn model for n=3
def knn_regressor(k_value, distance_metric, x3, y3):
  model_3d = KNeighborsRegressor(n_neighbors = k_value, algorithm='auto', leaf_size=30, metric='minkowski', p = distance_metric)
  model_3d.fit(x3, y3)
  return model_3d

k_values = [1,3,5,7]
distance_metric = [1,2]

for i in k_values:
  print("\nVALUE OF k", i)
  for k in distance_metric:
    model_3d = knn_regressor(i,k, x3, y3)
    #Calculating performance on training set
    if k == 1:
      print('Value of distance_metric: Manhattan')
    else:
      print('Value of distance_metric: Euclidean')

    prediction = model_3d.predict(x3)
    error = sqrt(mean_squared_error(y3,prediction))
    print('Root Mean Square Error (n = 3) :',error)

    prediction_last_week = model_3d.predict([[259,37, 27]])

    #final_result.append[]
    final_result.append(['3',i,k,error,round(prediction_last_week[0])])


VALUE OF k 1
Value of distance_metric: Manhattan
Root Mean Square Error (n = 3) : 0.0
Value of distance_metric: Euclidean
Root Mean Square Error (n = 3) : 0.0

VALUE OF k 3
Value of distance_metric: Manhattan
Root Mean Square Error (n = 3) : 9.319310384577744
Value of distance_metric: Euclidean
Root Mean Square Error (n = 3) : 9.421002927087079

VALUE OF k 5
Value of distance_metric: Manhattan
Root Mean Square Error (n = 3) : 10.390709708630459
Value of distance_metric: Euclidean
Root Mean Square Error (n = 3) : 10.618961089972512

VALUE OF k 7
Value of distance_metric: Manhattan
Root Mean Square Error (n = 3) : 10.764112415626471
Value of distance_metric: Euclidean
Root Mean Square Error (n = 3) : 10.835295886184086


In [ ]:
results = pd.DataFrame(final_result,columns=['Dimensions','K_Value','Distance_Metric','RMSE', 'Last_week_prediction'])
results.Distance_Metric = results.Distance_Metric.apply(lambda x: 'Manhattan' if x == 1 else 'Euclidean')
results['Actual_Value'] = 46
results['Difference in Real & Predicted Value'] = results['Actual_Value'] -  results['Last_week_prediction']
results

Dimensions  K_Value  ... Actual_Value  Difference in Real & Predicted Value
0           2        1  ...           46                                     0
1           2        1  ...           46                                     0
2           2        3  ...           46                                     2
3           2        3  ...           46                                     2
4           2        5  ...           46                                     4
5           2        5  ...           46                                     4
6           2        7  ...           46                                     4
7           2        7  ...           46                                     4
8           3        1  ...           46                                     0
9           3        1  ...           46                                     0
10          3        3  ...           46                                     2
11          3        3  ...           46                                     2
12          3        5  ...           46                                     4
13          3        5  ...           46                                     4
14          3        7  ...           46                                     4
15          3        7  ...           46                                     4

[16 rows x 7 columns]

Conclusion:

1. In case of k=1 iverfits the training data.
2. As k value is increasing the RMSE for the model is increasing.  
3. In conclusion as per the results we have obersed that the following combination of parameter where n = 2, k = 3, distance metric = Manhattan results into rmse = 8.923423 with given data. Results given are as follows:  Real value = 46 and Predicted Value = 44.